File: src/leetcode/m61/RotateList.java
/*
 * Copyright (c) 2024 THGEEK.COM. All rights reserved.
 */

package leetcode.m61;

import leetcode.ListNode;

/**
 * LeetCode 61
 *
 * @author Tao Hong
 * @version 1.0
 * @since 2024/10/03 08:15
 */
public class RotateList {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || k == 0) {
            return head;
        }

        int len = 0;
        ListNode fast = head;
        ListNode oldTail = null;
        while (fast != null) {
            len++;
            oldTail = fast;
            fast = fast.next;
        }
        if (k % len == 0) {
            return head;
        }
        int newHeadStart = len - k % len;

        oldTail.next = head;
        ListNode newHead = head;
        ListNode newTail;
        int i = 0;
        do {
            newTail = newHead;
            newHead = newHead.next;
            i++;
        } while (i < newHeadStart);
        newTail.next = null;

        return newHead;
    }
}

File: src/leetcode/m61/RotateList.go
package m61

import "leetcode/src/leetcode/support"

func rotateRight(head *support.ListNode, k int) *support.ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	listLen := 0
	curr := head
	var tail *support.ListNode
	for curr != nil {
		listLen++
		if curr.Next == nil {
			tail = curr
		}
		curr = curr.Next
	}

	k = k % listLen
	if k == 0 {
		return head
	}

	leftLen := listLen - k
	dummy := support.ListNode{Next: head}
	curr = &dummy
	for i := 0; i < leftLen; i++ {
		curr = curr.Next
	}

	newHead := curr.Next
	curr.Next = nil
	tail.Next = head
	return newHead
}
